presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+                        .map(Subfield.LongSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredStringKeys = null;
+                return;
+            case STRING:
+            case BINARY:
+                requiredStringKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.StringSubscript.class::cast)
+                        .map(Subfield.StringSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredLongKeys = null;
+                return;
+            default:
+                throw new IllegalArgumentException("Unsupported flat map key type: " + keyOrcType);
+        }
+    }
+
+    private static Optional<TupleDomainFilter> getTopLevelFilter(Map<Subfield, TupleDomainFilter> filters)
+    {
+        Map<Subfield, TupleDomainFilter> topLevelFilters = Maps.filterEntries(filters, entry -> entry.getKey().getPath().isEmpty());
+        if (topLevelFilters.isEmpty()) {
+            return Optional.empty();
+        }
+
+        checkArgument(topLevelFilters.size() == 1, "MAP column may have at most one top-level range filter");
+        TupleDomainFilter filter = Iterables.getOnlyElement(topLevelFilters.values());
+        checkArgument(filter == IS_NULL || filter == IS_NOT_NULL, "Top-level range filter on MAP column must be IS NULL or IS NOT NULL");
+        return Optional.of(filter);
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        allNulls = false;
+
+        if (outputRequired && nullsAllowed && presentStream != null) {
+            nulls = ensureCapacity(nulls, positionCount);
+        }
+
+        if (!nullsAllowed || !nonNullsAllowed) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
+        }
+        else {
+            outputPositions = positions;
+            outputPositionsReadOnly = true;
+        }
+
+        if (presentStream == null) {
+            readNoNulls(offset, positions, positionCount);
+        }
+        else if (keyCount == 0) {
+            readAllNulls(offset, positions, positionCount);
+        }
+        else {
+            readWithNulls(offset, positions, positionCount);
+        }
+
+        localMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        return outputPositionCount;
+    }
+
+    private void readNoNulls(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!nonNullsAllowed) {
+            outputPositionCount = 0;
+            return;
+        }
+
+        nestedLengths = ensureCapacity(nestedLengths, positionCount);
+        Arrays.fill(nestedLengths, 0);
```

Comment:
wondering when we will we need this, during initialization it will be 0, and when we are reusing the same index we can first set it to 0, instead of making the whole array fill at the start.

Reply:
@bhhari We need to initialize all values to zero because we are computing final values by adding one item at a time.

```
nestedLengths[i]++;
```